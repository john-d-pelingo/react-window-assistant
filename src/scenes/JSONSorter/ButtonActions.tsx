import { Box, Button, Popover, Typography } from '@material-ui/core'
import React, { FC, MouseEvent, useState } from 'react'
import styled from 'styled-components'

import { copyToClipboard } from 'helpers/copyToClipboard'

const StyledBox = styled(Box)`
  display: flex;
  margin-top: 24px;
  justify-content: flex-end;
`

const CopyButton = styled(Button)`
  && {
    margin-right: 24px;
  }
`

const StyledTypograhpy = styled(Typography)`
  padding: 16px;
`

interface ButtonActionsProps {
  canCopyJson: boolean
  canSort: boolean
  textToCopy?: string
}

export const ButtonActions: FC<ButtonActionsProps> = ({
  canCopyJson,
  canSort,
  textToCopy = '',
}) => {
  const [popOverAnchorEl, setPopOverAnchorEl] = useState<
    (EventTarget & HTMLButtonElement) | null
  >()

  const handleCopyClick = (event: MouseEvent<HTMLButtonElement>): void => {
    event.preventDefault()

    setPopOverAnchorEl(event.currentTarget)

    copyToClipboard(textToCopy)
  }

  const handlePopOverClose = (): void => {
    setPopOverAnchorEl(null)
  }

  const isPopOverOpen = Boolean(popOverAnchorEl)
  const id = isPopOverOpen ? 'simple-popover' : ''

  return (
    <StyledBox>
      <CopyButton
        aria-label="Copy button"
        aria-describedby={id}
        color="secondary"
        disabled={!canCopyJson}
        onClick={handleCopyClick}
        variant="contained"
      >
        Copy
      </CopyButton>
      <Popover
        anchorEl={popOverAnchorEl}
        anchorOrigin={{
          horizontal: 'left',
          vertical: 'center',
        }}
        id={id}
        open={isPopOverOpen}
        onClose={handlePopOverClose}
        transformOrigin={{
          horizontal: 'right',
          vertical: 'center',
        }}
      >
        <StyledTypograhpy>Copied JSON!</StyledTypograhpy>
      </Popover>
      <Button
        aria-label="Sort button"
        color="primary"
        disabled={!canSort}
        type="submit"
        variant="contained"
      >
        Sort
      </Button>
    </StyledBox>
  )
}
