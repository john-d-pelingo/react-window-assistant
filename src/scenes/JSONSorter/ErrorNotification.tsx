import { IconButton, Snackbar, SnackbarContent } from '@material-ui/core'
import { amber, blue, green, red } from '@material-ui/core/colors'
import { SnackbarContentProps } from '@material-ui/core/SnackbarContent'
import { SvgIconProps } from '@material-ui/core/SvgIcon'
import { CheckCircle, Close, Error, Info, Warning } from '@material-ui/icons'
import React, { ComponentType, FC } from 'react'
import styled from 'styled-components'

type Variant = 'error' | 'info' | 'success' | 'warning'

interface IErrorNotificationProps extends SnackbarContentProps {
  autoHideDuration?: number
  className?: string
  isSnackBarOpen?: boolean
  onClose: () => void
  variant?: Variant
}

type VariantIcon = { [key in Variant]: ComponentType<SvgIconProps> }

const variantIcon: VariantIcon = {
  error: Error,
  info: Info,
  success: CheckCircle,
  warning: Warning,
}

// eslint-disable-next-line
const StyledSnackbarContent = styled(({ colorVariant, ...otherProps }) => (
  <SnackbarContent {...otherProps} />
))<{
  colorVariant: Variant
}>`
  && {
    background-color: ${props => {
      switch (props.colorVariant) {
        case 'success':
          return green[600]
        case 'error':
          return red[600]
        case 'warning':
          return amber[700]
        case 'info':
        default:
          return blue[600]
      }
    }};
  }
`

const StyledClose = styled(Close)`
  font-size: 20px;
`

const Message = styled.span`
  display: flex;
  align-items: center;
`

export const ErrorNotification: FC<IErrorNotificationProps> = ({
  autoHideDuration = 6000,
  className = '',
  isSnackBarOpen = false,
  message,
  onClose,
  variant = 'success',
  ...snackbarContentProps
}) => {
  const Icon = variantIcon[variant]

  const StyledIcon = styled(Icon)`
    font-size: 20px;
    margin-right: 8px;
    opacity: 0.9;
  `

  return (
    <Snackbar
      anchorOrigin={{
        horizontal: 'left',
        vertical: 'bottom',
      }}
      autoHideDuration={6000}
      onClose={() => 1}
      open={isSnackBarOpen}
    >
      <StyledSnackbarContent
        action={[
          <IconButton
            aria-label="Close error notification"
            color="inherit"
            key="close"
            onClick={onClose}
          >
            <StyledClose />
          </IconButton>,
        ]}
        aria-describedby="client-snackbar"
        className={className}
        message={
          <Message id="client-snackbar">
            <StyledIcon />
            {message}
          </Message>
        }
        colorVariant={variant}
        {...snackbarContentProps}
      />
    </Snackbar>
  )
}
