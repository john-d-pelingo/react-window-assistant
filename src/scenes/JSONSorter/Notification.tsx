import { IconButton, Snackbar, SnackbarContent } from '@material-ui/core'
import { SnackbarContentProps } from '@material-ui/core/SnackbarContent'
import { SvgIconProps } from '@material-ui/core/SvgIcon'
import { amber, blue, green, red } from '@material-ui/core/colors'
import { CheckCircle, Close, Error, Info, Warning } from '@material-ui/icons'
import { ComponentType, FC } from 'react'
import styled from 'styled-components'

type Category = 'error' | 'info' | 'success' | 'warning'

interface NotificationProps extends SnackbarContentProps {
  autoHideDuration?: number
  category?: Category
  className?: string
  isSnackBarOpen?: boolean
  onClose: () => void
}

type CategoryIcon = { [key in Category]: ComponentType<SvgIconProps> }

const categoryIcon: CategoryIcon = {
  error: Error,
  info: Info,
  success: CheckCircle,
  warning: Warning,
}

// eslint-disable-next-line
const StyledSnackbarContent = styled(({ category, ...otherProps }) => (
  <SnackbarContent {...otherProps} />
))<{
  category: Category
}>`
  && {
    background-color: ${(props) => {
      switch (props.category) {
        case 'success':
          return green[600]
        case 'error':
          return red[600]
        case 'warning':
          return amber[700]
        case 'info':
        default:
          return blue[600]
      }
    }};
  }
`

const StyledClose = styled(Close)`
  font-size: 20px;
`

const Message = styled.span`
  display: flex;
  align-items: center;
`

export const Notification: FC<NotificationProps> = ({
  category = 'success',
  className = '',
  isSnackBarOpen = false,
  message,
  onClose,
  ...snackbarContentProps
}) => {
  const Icon = categoryIcon[category]

  const StyledIcon = styled(Icon)`
    font-size: 20px;
    margin-right: 8px;
    opacity: 0.9;
  `

  const StyledIconComponent = () => <StyledIcon />

  return (
    <Snackbar
      anchorOrigin={{
        horizontal: 'left',
        vertical: 'bottom',
      }}
      autoHideDuration={6000}
      onClose={() => 1}
      open={isSnackBarOpen}
    >
      <StyledSnackbarContent
        action={[
          <IconButton
            aria-label="Close error notification"
            color="inherit"
            key="close"
            onClick={onClose}
          >
            <StyledClose />
          </IconButton>,
        ]}
        aria-describedby="client-snackbar"
        category={category}
        className={className}
        message={
          <Message id="client-snackbar">
            <StyledIconComponent />
            {message}
          </Message>
        }
        {...snackbarContentProps}
      />
    </Snackbar>
  )
}
