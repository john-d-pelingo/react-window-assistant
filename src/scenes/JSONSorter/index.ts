export { JSONSorter as default } from './JSONSorter'
