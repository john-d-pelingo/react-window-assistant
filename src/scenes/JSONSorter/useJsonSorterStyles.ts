import { createStyles, makeStyles } from '@material-ui/core'

export const useJsonSorterStyles = makeStyles(theme =>
  createStyles({
    container: {
      margin: `${theme.spacing(12.5)}px 0`,
    },
    text: {
      fontFamily: ['Courier New', 'sans-serif'].join(','),
      fontSize: '20px',
    },
    submitBox: {
      marginTop: `${theme.spacing(2)}px`,
    },
    submit: {
      float: 'right',
    },
  }),
)
