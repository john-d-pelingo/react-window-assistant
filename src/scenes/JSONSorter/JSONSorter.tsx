import { Container, OutlinedInput } from '@material-ui/core'
import { RouteComponentProps } from '@reach/router'
import React, {
  ChangeEvent,
  createRef,
  FC,
  FormEvent,
  KeyboardEvent,
  useEffect,
  useRef,
  useState,
} from 'react'
import sortKeys from 'sort-keys'

import { ButtonActions } from './ButtonActions'
import { ErrorNotification } from './ErrorNotification'
import { useJsonSorterStyles } from './useJsonSorterStyles'

export const JSONSorter: FC<RouteComponentProps> = () => {
  const classes = useJsonSorterStyles()
  const inputNode = useRef<HTMLTextAreaElement>()
  const formNode = createRef<HTMLFormElement>()
  const [jsonText, setJsonText] = useState('')
  const [canCopyJson, setCanCopyJson] = useState(false)
  const [isErrorNotificationOpen, setIsErrorNotificationOpen] = useState(false)

  useEffect(() => {
    if (inputNode.current) {
      inputNode.current.focus()
    }
  }, [])

  const handleInputChange = (event: ChangeEvent<HTMLTextAreaElement>): void => {
    const { value } = event.target

    setJsonText(value)

    try {
      const parsedText = JSON.parse(value)
      sortKeys(parsedText)

      setCanCopyJson(true)
    } catch (error) {
      setCanCopyJson(false)
    }
  }

  const handleInputKeyDown = (
    event: KeyboardEvent<HTMLTextAreaElement>,
  ): void => {
    if (isErrorNotificationOpen) {
      setIsErrorNotificationOpen(false)
    }

    if (
      (event.ctrlKey || event.metaKey) &&
      (event.keyCode === 13 || event.keyCode === 10) &&
      formNode.current
    ) {
      handleSubmit()
    }
  }

  const handleSubmit = (event?: FormEvent<HTMLFormElement>): void => {
    if (event) {
      event.preventDefault()
    }

    try {
      const parsedText = JSON.parse(jsonText)
      const sortedKeys = sortKeys(parsedText)

      setJsonText(JSON.stringify(sortedKeys, undefined, 2) + '\n')
      setCanCopyJson(true)

      if (inputNode.current) {
        inputNode.current.focus()
      }
    } catch (error) {
      setCanCopyJson(false)
      setIsErrorNotificationOpen(true)

      if (inputNode.current) {
        inputNode.current.focus()
      }
    }
  }

  return (
    <Container className={classes.container} maxWidth="xl">
      <ErrorNotification
        autoHideDuration={5000}
        isSnackBarOpen={isErrorNotificationOpen}
        message="Invalid JSON!"
        onClose={() => {
          setIsErrorNotificationOpen(false)
        }}
        variant="error"
      />
      <form onSubmit={handleSubmit} ref={formNode}>
        <OutlinedInput
          className={classes.inputText}
          fullWidth
          inputProps={{
            'aria-label': 'JSON input',
          }}
          inputRef={inputNode}
          labelWidth={0}
          multiline
          onChange={handleInputChange}
          onKeyDown={handleInputKeyDown}
          placeholder="Paste your JSON object here"
          rows={20}
          value={jsonText}
        />
        <ButtonActions
          canCopyJson={canCopyJson}
          canSort={jsonText.trim().length !== 0}
          textToCopy={jsonText}
        />
      </form>
    </Container>
  )
}
