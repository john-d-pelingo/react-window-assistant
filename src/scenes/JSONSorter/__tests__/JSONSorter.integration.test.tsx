import 'jest-dom/extend-expect'

import React from 'react'
import { fireEvent, render } from 'react-testing-library'

import { JSONSorter } from '../JSONSorter'

describe('scenes - JSONSorter', () => {
  beforeEach(() => {
    document.getSelection = jest.fn()
    document.execCommand = jest.fn()
  })

  afterEach(() => {
    jest.clearAllMocks()
  })

  it('allows user to click sort but not copy when the json is invalid', () => {
    const { getByLabelText, getByText, queryByText } = render(<JSONSorter />)

    const jsonInputElement = getByLabelText('JSON input') as HTMLInputElement

    expect(document.activeElement).toEqual(jsonInputElement)

    const copyButtonElement = getByLabelText('Copy button') as HTMLButtonElement
    const sortButtonElement = getByLabelText('Sort button') as HTMLButtonElement

    expect(copyButtonElement.disabled).toBe(true)
    expect(sortButtonElement.disabled).toBe(true)

    fireEvent.change(jsonInputElement, {
      target: {
        value: 'bad json',
      },
    })

    expect(jsonInputElement).toHaveTextContent('bad json')
    expect(copyButtonElement.disabled).toBe(true)
    expect(sortButtonElement.disabled).toBe(false)

    expect(queryByText('Invalid JSON!')).not.toBeInTheDocument()

    fireEvent.click(sortButtonElement)

    const errorNotificationElement = getByText('Invalid JSON!')

    expect(errorNotificationElement).toBeInTheDocument()
  })

  it('copies the json when the user clicks sort and copy', () => {
    const { getByLabelText, queryByText } = render(<JSONSorter />)

    const jsonInputElement = getByLabelText('JSON input') as HTMLInputElement
    const copyButtonElement = getByLabelText('Copy button') as HTMLButtonElement
    const sortButtonElement = getByLabelText('Sort button') as HTMLButtonElement

    fireEvent.change(jsonInputElement, {
      target: {
        value: '{"z":2,"y":0,"m":{"gg":42,"become":"you"},"p":1}',
      },
    })

    expect(jsonInputElement).toHaveTextContent(
      '{"z":2,"y":0,"m":{"gg":42,"become":"you"},"p":1}',
    )
    expect(copyButtonElement.disabled).toBe(false)
    expect(sortButtonElement.disabled).toBe(false)

    fireEvent.click(sortButtonElement)

    expect(queryByText('Invalid JSON!')).not.toBeInTheDocument()
    expect(jsonInputElement).toHaveTextContent(
      '{ "m": { "gg": 42, "become": "you" }, "p": 1, "y": 0, "z": 2 }',
    )

    fireEvent.click(copyButtonElement)

    expect(queryByText('Copied JSON!')).toBeInTheDocument()
  })

  it('does not sort the JSON input when it is invalid and user presses ctrl-enter and cmd-center', () => {
    const { getByLabelText, queryByText } = render(<JSONSorter />)

    const jsonInputElement = getByLabelText('JSON input') as HTMLInputElement

    fireEvent.change(jsonInputElement, {
      target: {
        value: 'nemesis',
      },
    })

    fireEvent.keyDown(jsonInputElement, {
      ctrlKey: true,
      keyCode: 13,
    })

    expect(queryByText('Invalid JSON!')).toBeInTheDocument()

    const closeErrorNotificationElement = getByLabelText(
      'Close error notification',
    )

    fireEvent.click(closeErrorNotificationElement)
  })

  it('does not sort the JSON input when it is invalid and user presses ctrl-enter', () => {
    const { getByLabelText, queryByText } = render(<JSONSorter />)

    const jsonInputElement = getByLabelText('JSON input') as HTMLInputElement

    fireEvent.change(jsonInputElement, {
      target: {
        value: 'Am I valid?',
      },
    })

    fireEvent.keyDown(jsonInputElement, {
      ctrlKey: true,
      keyCode: 13,
    })

    expect(queryByText('Invalid JSON!')).toBeInTheDocument()

    const closeErrorNotificationElement = getByLabelText(
      'Close error notification',
    )

    fireEvent.click(closeErrorNotificationElement)
  })

  it('does not sort the JSON input when it is invalid and user presses ctrl-enter', () => {
    const { getByLabelText, queryByText } = render(<JSONSorter />)

    const jsonInputElement = getByLabelText('JSON input') as HTMLInputElement

    fireEvent.change(jsonInputElement, {
      target: {
        value: 'Not really ...',
      },
    })

    fireEvent.keyDown(jsonInputElement, {
      metaKey: true,
      keyCode: 10,
    })

    expect(queryByText('Invalid JSON!')).toBeInTheDocument()

    const closeErrorNotificationElement = getByLabelText(
      'Close error notification',
    )

    fireEvent.click(closeErrorNotificationElement)
  })

  it('sorts the JSON input when it valid and user presses ctrl-enter', () => {
    const { getByLabelText } = render(<JSONSorter />)

    const jsonInputElement = getByLabelText('JSON input') as HTMLInputElement

    fireEvent.change(jsonInputElement, {
      target: {
        value: '{"l": 4,            ".": [2, 3], "j": "k"}',
      },
    })

    fireEvent.keyDown(jsonInputElement, {
      metaKey: true,
      keyCode: 10,
    })

    expect(jsonInputElement.textContent).toBe(
      '{\n' +
        '  ".": [\n' +
        '    2,\n' +
        '    3\n' +
        '  ],\n' +
        '  "j": "k",\n' +
        '  "l": 4\n' +
        '}\n',
    )
  })
})
