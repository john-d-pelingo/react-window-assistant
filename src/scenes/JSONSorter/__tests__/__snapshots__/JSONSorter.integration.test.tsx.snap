// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`scenes - JSONSorter mounts 1`] = `
.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-top: 24px;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
}

.c2.c2 {
  margin-right: 24px;
}

.c0 {
  font-family: 'Courier New',mono,sans-serif;
  font-size: 16px;
}

<div
  class="MuiContainer-root MuiContainer-maxWidthXl"
  style="margin: 100px 0px;"
>
  <form>
    <div
      class="MuiInputBase-root MuiOutlinedInput-root c0 MuiInputBase-fullWidth Mui-focused Mui-focused MuiInputBase-multiline MuiOutlinedInput-multiline"
    >
      <textarea
        aria-label="JSON input"
        class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline MuiOutlinedInput-inputMultiline"
        placeholder="Paste your JSON object here"
        rows="20"
      />
      <fieldset
        aria-hidden="true"
        class="PrivateNotchedOutline-root-48 MuiOutlinedInput-notchedOutline"
        style="padding-left: 8px;"
      >
        <legend
          class="PrivateNotchedOutline-legend-49"
          style="width: 0px;"
        >
          <span>
            â€‹
          </span>
        </legend>
      </fieldset>
    </div>
    <div
      class="MuiBox-root MuiBox-root-51 c1"
    >
      <button
        aria-describedby=""
        aria-label="Copy button"
        class="MuiButtonBase-root MuiButton-root c2 MuiButton-contained MuiButton-containedSecondary Mui-disabled Mui-disabled"
        disabled=""
        tabindex="-1"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          Copy
        </span>
      </button>
      <button
        aria-label="Sort button"
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary Mui-disabled Mui-disabled"
        disabled=""
        tabindex="-1"
        type="submit"
      >
        <span
          class="MuiButton-label"
        >
          Sort
        </span>
      </button>
    </div>
  </form>
</div>
`;
